### ISSUE TITLE
[Discussion] Youtube bestvideo and vp9 preference.

### ISSUE THREAD

**Issue by xioren** [yt-dlp/yt-dlp#1232]
## Checklist

- [x] I'm asking a question
- [x] I've looked through the README
- [x] I've read the opening an issue section in CONTRIBUTING.md
- [x] I've searched the bugtracker for similar questions including closed ones
- [x] I have given an appropriate title to the issue

The bestvideo format option does not select the best video stream in general, instead it seems to get the best vp9 stream because of the vp9 preference of yt-dlp. There are situations where the best vp9 stream in non trivially worse than the best h.264 stream. An example: 

```
299 mp4  1920x1080  50  |  249.17MiB 3645k https | avc1.64002a 3645k                   1080p50, mp4_dash
303 webm 1920x1080  50  |  136.62MiB 1998k https | vp9         1998k                   1080p50, webm_dash
```

In the above example yt-dlp will select itag 303 which is nearly half the bitrate of 299. In comparing screenshots of the two streams, the vp9 stream is noticeably lower quality.

"beststream" seems like a misnomer and can lead users to believe they are in fact downloading the best stream when they aren't. Is this the best approach? Is there a reliable way to select the best stream irrespective of codec?

**Comment by xioren** [comment]
> define `best`

clearest picture (less compression artifacts) at highest resolution

**Comment by Ashish0804** [comment]
Okay so you are suggesting we ship a Neural Network with dlp and then download all formats (partially maybe), then check which one is `best` and then download it completely for all videos?

**Comment by xioren** [comment]
Snarky responses notwithstanding, I think my description of best was quite universal; one that yt-dlp itself adheres when choosing the "best" vp9 stream. It is trivial to compare streams of the same codec and choose the best quality, the difficulty clearly is with comparing streams of different codecs to which I asked if there was a reliable way to do so.

**Comment by pukkandan** [comment]
This is a complicated issue. Firstly, "best" is a very subjective term. Most often, the formats have very little difference and which one provides better quality is not objectively determinable - that is, even if we had access to both files and the source (not to mention that we dont have access to the source). Even when there is a significant perceivable difference in quality we need to be able to determine that (1) in an automated way (2) without downloading the said videos and (3) with the often limted metadata that the website provides.

So, we decide the `bestvideo` based on a list of pre-configured parameters. **youtube-dl** considers the `tbr` (total bitrate) to the the most significant factor in determining a video's quality. However, since modern codecs are generally more efficient than older ones, this is not an accurate measure of quality.

In yt-dlp, I made the *decision* to prioritize the resolution and codec over bitrate. This is also ofc not an objective measure of quality - it was just my personal decision based on my experiences with downloading youtube videos. The decision was based on the logic that most websites (including YouTube) encode their videos in different codecs so that they can serve various devices that may/maynot support them. As such, they **generally try** to have all the formats (given the same resolution) to be of a similar *perceived* quality. Therefore, it is generally desirable to download the format with the best codec.

I understand that not everyone will agree to this definition of "best". So rather than simply changing the sorting order, I also added the `--sort-formats` option so that each user can determine for themselves what factors should take precedence in determining the best format. You can read the extensive documentation of this option and of the default behavior in https://github.com/yt-dlp/yt-dlp#sorting-formats

Also, note the use of "generally" in the above paragraphs. This is the root cause of the problem you are facing. The video you are pointing out is probably (I dont know for sure, since u didnt give link) an edge case where our blanket assumptions about the quality does not hold. 

I don't know how we can improve our sorting mechanism to compare b/w the various codecs. If you have any ideas, I'd like to hear them. But understand that whatever we do, we will only be at best slightly improving the situation and there will **always** be edge cases. There really is nothing we can do about that.

**tldr**:
1. "best" is subjective
2. Even when there is a significant differences for it to be not subjective, that cannot be determined by available metadata
3. youtube-dl chooses bestvideo based on bitrate while yt-dlp instead looks at resolution and codec
5. If you don't agree with the default order, you can use `-S` to set a custom order
6. Whatever we do, there will be edge cases were the default selection is best. Your only option is to manually download and compare both formats in that case
7. Even so, if you have suggestions for how to improve the current defaults, I am all ears

**Comment by xioren** [comment]
Thanks for the well thought out response. I will concede that "best" is subjective to an extent, but I also posit that most people will agree that a 1080p stream with a 1.906 MiB/s bit-rate is "better" than a 480p stream with a 403.389 KiB/s bit-rate. Admittedly this type of comparison breaks down when the differences between streams are more subtle (i.e. same resolution and similar bit-rates).

I have been searching all day for ways to compare vp9 and h.264 without much luck. Ideally you would want a way to determine a ratio between the bit-rates. i.e. a vp9 stream of n bit-rate is roughly equivalent (visually imperceptible) to an h.264 stream of k bit-rate; so given any stream of either format, you could apply the ratio to obtain roughly what the bit-rate of the other format should be for the same visual quality. Maybe that is a pipe dream and as there doesn't seem to be any established way to do so at this time I will close the issue.

**Comment by alakani** [comment]
> > Ideally you would want a way to determine a ratio between the bit-rates. i.e. a vp9 stream of n bit-rate is roughly equivalent (visually imperceptible) to an h.264 stream of k bit-rate; so given any stream of either format, you could apply the ratio to obtain roughly what the bit-rate of the other format should be for the same visual quality. Maybe that is a pipe dream and as there doesn't seem to be any established way to do so at this time
> 
> This has been proposed before. If we can agree on a weight for each codec, implementing a weighted comparison in code is trivial. But as you mentioned, I have not found any established metrics to do such a comparison. If you manage to find any such standards or reputable studies for such a metric, let me know and we can consider using it (if not by default, then atleast as an option)

I was going to try this in 2019 but got dstftw'd; he said he would throw out the PR and laugh if I touched his precious switch spaghetti. No need to include the NN with the release, can precompute MOS weights for the various codecs using Netflix's VMAS (if it hasn't been trained yet for YT codecs I have dozens of TB where I manually picked the best codec). It would only be 90-96% accurate since codec quality sometimes depends on specific video content (some codecs do better with high motion vs high detail etc) - but it could take a pretty good guess, and output a warning if the user might want to download 2 formats and compare manually if quality is that critical. Or OP's idea about vast filesize differences is decent too and much simpler to implement for the more obvious cases. Or could crowdsource it like sponsorblock. All that said I don't think it's as big of a deal as I did back then, due to changes on YT's end making the issue less frequent combined with sane defaults in yt-dlp. Either way thank you so much for the fork, I wish I found it months ago :)

**Comment by dreamflasher** [comment]
From the verbose output I got the information that "bestvideo" is currently defined as `hasvid, ie_pref, quality, res, fps, hdr:12(7), source, lang, vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, id`.

So, do I understand correctly if there is a newer coded at a smaller resolution, then this is currently preferred?
If I wanted to prefer higher resolution, and then select for newer codecs I would need to pass:
`--format-sort res,quality,fps,hdr:12(7),source,lang,vcodec:vp9.2(10),channels,acodec,filesize,fs_approx,tbr,vbr,abr,asr,proto,vext,aext,hasaud,id`
Correct?

**Comment by pukkandan** [comment]
> ..., res, ..., vcodec:vp9.2(10), ...

> So, do I understand correctly if there is a newer coded at a smaller resolution, then this is currently preferred?

No. `res` has higher priority than `vcodec`

**Comment by dreamflasher** [comment]
What does `quality` mean then? The documentation says ` The quality of the format`? And by default quality is before res?

**Comment by pukkandan** [comment]
The extractors can report some formats to be of a better quality than others irrespective of other fields. For example, premium aac audio on youtube (-f141) is considered to be a higher "quality" than the webm (-f251). This ensures they are sorted correctly despite the other indicators saying otherwise

**Comment by dreamflasher** [comment]
Okay, but shouldn't the `res` have the higher priority? Even if there's an extractor reporting `premium video at 720p`, and there's a `mp4 video at 1000p` I would prefer the latter?

**Comment by chrizilla** [comment]
@dreamflasher : very good question! 👍 
I have been wondering about this, too. Hope we'll get a response.

**Comment by pukkandan** [comment]
No, it was designed to be higher priority. In case the extractor wants it beelow resolution, it can change the field order.

**Comment by MoonLord-LM** [comment]
> This is a complicated issue. Firstly, "best" is a very subjective term. Most often, the formats have very little difference and which one provides better quality is not objectively determinable - that is, even if we had access to both files and the source (not to mention that we dont have access to the source). Even when there is a significant perceivable difference in quality we need to be able to determine that (1) in an automated way (2) without downloading the said videos and (3) with the often limted metadata that the website provides.
> 
> So, we decide the `bestvideo` based on a list of pre-configured parameters. **youtube-dl** considers the `tbr` (total bitrate) to the the most significant factor in determining a video's quality. However, since modern codecs are generally more efficient than older ones, this is not an accurate measure of quality.
> 
> In yt-dlp, I made the _decision_ to prioritize the resolution and codec over bitrate. This is also ofc not an objective measure of quality - it was just my personal decision based on my experiences with downloading youtube videos. The decision was based on the logic that most websites (including YouTube) encode their videos in different codecs so that they can serve various devices that may/maynot support them. As such, they **generally try** to have all the formats (given the same resolution) to be of a similar _perceived_ quality. Therefore, it is generally desirable to download the format with the best codec.
> 
> I understand that not everyone will agree to this definition of "best". So rather than simply changing the sorting order, I also added the `--sort-formats` option so that each user can determine for themselves what factors should take precedence in determining the best format. You can read the extensive documentation of this option and of the default behavior in https://github.com/yt-dlp/yt-dlp#sorting-formats
> 
> Also, note the use of "generally" in the above paragraphs. This is the root cause of the problem you are facing. The video you are pointing out is probably (I dont know for sure, since u didnt give link) an edge case where our blanket assumptions about the quality does not hold.
> 
> I don't know how we can improve our sorting mechanism to compare b/w the various codecs. If you have any ideas, I'd like to hear them. But understand that whatever we do, we will only be at best slightly improving the situation and there will **always** be edge cases. There really is nothing we can do about that.
> 
> **tldr**:
> 
> 1. "best" is subjective
> 2. Even when there is a significant differences for it to be not subjective, that cannot be determined by available metadata
> 3. youtube-dl chooses bestvideo based on bitrate while yt-dlp instead looks at resolution and codec
> 4. If you don't agree with the default order, you can use `-S` to set a custom order
> 5. Whatever we do, there will be edge cases were the default selection isn't best. Your only option is to manually download and compare both formats in that case
> 6. Even so, if you have suggestions for how to improve the current defaults, I am all ears

This issue is indeed a bit complex.  
I downloaded both a smaller-sized VP9 file and a larger AVC1 file, and honestly, I can't tell which one is better.

**Comment by yommish** [comment]
> What does `quality` mean then? The documentation says ` The quality of the format`? And by default quality is before res?

For anyone like me who required more detail, you can get a more tangible idea by looking at the extractor for YouTube specifically lines 3160-3319 https://github.com/yt-dlp/yt-dlp/blob/26feac3dd142536ad08ad1ed731378cb88e63602/yt_dlp/extractor/youtube/_video.py#L3160

Also see https://github.com/yt-dlp/yt-dlp/blob/26feac3dd142536ad08ad1ed731378cb88e63602/yt_dlp/utils/_utils.py#L2829

Video files are classified as one of: 'tiny', 'small', 'medium', 'large', 'hd720', 'hd1080', 'hd1440', 'hd2160', 'hd2880', 'highres' which is then turned into a numeric value…

Glad to be slightly less confused about quality. Was feeling a bit like Robert Pirsig.

**Comment by github-account1111** [comment]
From [r/youtubedl]:

> - For 60FPS videos, AVC (299) is MUCH better than VP9 (303, yt-dlp default)
> - For 30FPS videos, VP9 Premium (616) is better than AVC. But when premium bitrate is not available, AVC is better than VP9.

[r/youtubedl]: https://www.reddit.com/r/youtubedl/comments/17j4ozo/want_the_best_quality_files_you_might_want_to
